cpp/src/arrow/filesystem/hdfs_test.cc

```diff
@@ -0,0 +1,169 @@
+      use_hdfs3 = true;
+    }
+
+    HdfsOptions hdfs_options;
+    hdfs_options.ConfigureEndPoint(hdfs_host, hdfs_port);
+    hdfs_options.ConfigureHdfsUser(hdfs_user);
+    hdfs_options.ConfigureHdfsDriver(use_hdfs3);
+    hdfs_options.ConfigureHdfsReplication(0);
+
+    std::shared_ptr<HadoopFileSystem> hdfs;
+    auto status = HadoopFileSystem::Make(hdfs_options, &hdfs);
+    if (!status.ok()) {
+      loaded_driver_ = false;
+      return;
+    }
+    loaded_driver_ = true;
+    fs_ = std::make_shared<SubTreeFileSystem>("", hdfs);
+  }
+
+ protected:
+  std::shared_ptr<FileSystem> fs_;
+  bool loaded_driver_ = false;
+};
+
+HdfsDriver JNIDriver::type = HdfsDriver::LIBHDFS;
+HdfsDriver PivotalDriver::type = HdfsDriver::LIBHDFS3;
+
+typedef ::testing::Types<JNIDriver, PivotalDriver> DriverTypes;
+
+TYPED_TEST_CASE(TestHadoopFileSystem, DriverTypes);
+
+#define SKIP_IF_NO_DRIVER()                                  \
+  if (!this->loaded_driver_) {                               \
+    std::cout << "Driver not loaded, skipping" << std::endl; \
+    return;                                                  \
+  }
+
+TYPED_TEST(TestHadoopFileSystem, CreateDir) {
+  SKIP_IF_NO_DRIVER();
+
+  std::string directory_name = "AB";
+  ASSERT_OK(this->fs_->CreateDir(directory_name));
+}
+
+TYPED_TEST(TestHadoopFileSystem, GetTargetStats) {
+  SKIP_IF_NO_DRIVER();
+
+  std::string directory_name = "AB";
+  FileStats stat;
+  ASSERT_OK(this->fs_->GetTargetStats(directory_name, &stat));
+  AssertFileStats(stat, "AB", FileType::Directory);
+}
+
+TYPED_TEST(TestHadoopFileSystem, MoveDir) {
+  SKIP_IF_NO_DRIVER();
+
+  std::string directory_name_src = "AB";
+  std::string directory_name_dest = "CD";
+  ASSERT_OK(this->fs_->Move(directory_name_src, directory_name_dest));
+  FileStats stat;
+  ASSERT_OK(this->fs_->GetTargetStats(directory_name_dest, &stat));
+  AssertFileStats(stat, "CD", FileType::Directory);
+}
+
+TYPED_TEST(TestHadoopFileSystem, WriteReadFile) {
+  SKIP_IF_NO_DRIVER();
+
+  std::string file_name = "CD/abc";
+  std::string data = "some data";
+  std::shared_ptr<io::OutputStream> stream;
+  ASSERT_OK(this->fs_->OpenOutputStream(file_name, &stream));
+  auto data_size = static_cast<int64_t>(data.size());
+  ASSERT_OK(stream->Write(data.data(), data_size));
+  ASSERT_OK(stream->Close());
+
+  std::shared_ptr<io::RandomAccessFile> file;
+  ASSERT_OK(this->fs_->OpenInputFile(file_name, &file));
+  int64_t file_size;
+  ASSERT_OK(file->GetSize(&file_size));
+  ASSERT_EQ(SOME_DATA_SIZE, file_size);
+  uint8_t buffer[SOME_DATA_SIZE];
+  int64_t bytes_read = 0;
+  ASSERT_OK(file->Read(SOME_DATA_SIZE, &bytes_read, buffer));
+  ASSERT_EQ(0, std::memcmp(buffer, data.c_str(), SOME_DATA_SIZE));
+}
+
+TYPED_TEST(TestHadoopFileSystem, DeleteFile) {
+  SKIP_IF_NO_DRIVER();
+
+  std::string file_name = "CD/abc";
+  ASSERT_OK(this->fs_->DeleteFile(file_name));
+  FileStats stat;
+  ASSERT_TRUE(Status::OK() != this->fs_->GetTargetStats(file_name, &stat));
```

Comment:
Here `GetTargetStats` should return ok and the file type would be `FileType::NonExistent`.